// @testim/testim-cli/commons/testimServicesApi.js
'use strict';

const pako = require('pako');
const _ = require('lodash');
const testimCustomToken = require('./testimCustomToken');
const constants = require('./constants');
const Promise = require('bluebird');
const utils = require('../utils.js');
const config = require('./config');
const httpRequest = require('./httpRequest');

const runnerVersion = utils.getRunnerVersion();
const logger = require('./logger').getLogger('testim service api');
const hash = require('object-hash');
const { ArgError } = require('../errors');

const DEFAULT_REQUEST_RETRY = 3;

function getTokenHeader() {
    return testimCustomToken.getCustomTokenV3()
        .then(brearToken => {
            if (!brearToken) {
                return Promise.reject(new Error('Failed to get token from server'));
            }
            return { Authorization: `Bearer ${brearToken}` };
        });
}

function postAuth({
    url, body, headers = {}, timeout, retry,
}) {
    return getTokenHeader()
        .then(tokenHeaders => {
            const finalHeaders = Object.assign({}, headers, tokenHeaders);
            return httpRequest.post({
                url: `${config.SERVICES_HOST}${url || ''}`,
                body,
                headers: finalHeaders,
                timeout,
                retry,
            });
        });
}

function postAuthFormData(url, fields, files, headers = {}, timeout) {
    return getTokenHeader()
        .then(tokenHeaders => {
            const finalHeaders = Object.assign({}, headers, tokenHeaders);
            return httpRequest.postForm(`${config.SERVICES_HOST}${url || ''}`, fields, files, finalHeaders, timeout);
        });
}

function putAuth(url, body) {
    return getTokenHeader()
        .then(headers => httpRequest.put(`${config.SERVICES_HOST}${url || ''}`, body, headers));
}

function getWithAuth(url, query, options, timeout) {
    return getTokenHeader()
        .then(headers => httpRequest.get(`${config.SERVICES_HOST}${url || ''}`, query, headers, timeout, options));
}

function getS3Artifact(url, timeout) {
    return utils.runWithRetries(() => getWithAuth(`/storage${url}`, null, { isBinary: true }, timeout));
}

function getTestPlan(projectId, testPlanNames) {
    //TODO: need to be checked after 3 months to prevent users from using old version
    const parseProperty = (dataValue) => (dataValue == null ? [] : (typeof (dataValue) === 'string' && JSON.parse(dataValue)) || dataValue);
    return utils.runWithRetries(() => getWithAuth('/testPlan', { projectId, name: testPlanNames.join(',') }))
        .then(body => body.map(testPlan => {
            testPlan.testConfigIds = parseProperty(testPlan.testConfigIds);
            testPlan.beforeAllLabels = parseProperty(testPlan.beforeAllLabels);
            testPlan.testLabels = parseProperty(testPlan.testLabels);
            testPlan.afterAllLabels = parseProperty(testPlan.afterAllLabels);
            return testPlan;
        }));
}


function loadTest({ testId, branch, projectId, skipSharedSteps = false, useBranchMap = true }) {
    return utils.runWithRetries(() => getWithAuth(`/test/${testId}`, {
        projectId,
        branch,
        skipSharedSteps,
        useBranchMap,
    }));
}

function saveTestPlanResult(projectId, testPlanId, result) {
    return utils.runWithRetries(() => postAuth({ url: '/testPlan/result', body: { projectId, testPlanId, result } }));
}

function updateTestStatus(projectId, executionId, testId, resultId, status, data, apiRetries = DEFAULT_REQUEST_RETRY) {
    return utils.runWithRetries(() => putAuth('/result/run/test', {
        runId: executionId,
        testId,
        resultId,
        status,
        projectId,
        runnerVersion,
        ...data,
    }), apiRetries);
}

function updateExecutionTests(executionId, runnerStatuses, status, reason, success, startTime, endTime, projectId) {
    return utils.runWithRetries(() => putAuth('/result/run/tests', {
        runId: executionId,
        runnerStatuses,
        status,
        reason,
        success,
        startTime,
        endTime,
        projectId,
    }), DEFAULT_REQUEST_RETRY);
}

function reportExecutionStarted({
    executionId,
    projectId,
    labels,
    startTime,
    executions,
    config,
    resultLabels,
    remoteRunId,
    localRunUserId,
    isLocalRun,
}) {
    const isCiRun = require('../cli/isCiRun').isCi;
    return postAuth({
        timeout: 90000,
        url: '/result/run',
        body: {
            runId: executionId,
            projectId,
            labels,
            startTime,
            execution: executions,
            status: 'RUNNING',
            config,
            resultLabels,
            remoteRunId,
            metadata: {
                isCiRun,
                localRunUserId,
                isLocalRun,
            },
        },
        retry: 3, // TODO: add a log in the callback
    });
}

function reportExecutionFinished(status, executionId, projectId, success, tmsOptions = {}, remoteRunId, resultExtraData) {
    const endTime = Date.now();

    return utils.runWithRetries(() => putAuth('/result/run', {
        runId: executionId,
        projectId,
        endTime,
        status,
        success,
        tmsOptions,
        remoteRunId,
        resultExtraData,
    }), DEFAULT_REQUEST_RETRY);
}

function getTestPlanTestList(projectId, names, planIds, branch) {
    return utils.runWithRetries(() => postAuth({
        url: '/testPlan/list',
        body: { projectId, names, planIds, branch },
        // people who send insane lists get a timeout :(
        timeout: 120000,
    }));
}

function getSuiteTestList({
    projectId, labels, testIds, testNames, testConfigNames, suiteNames, suiteIds, branch, rerunFailedByRunId, testConfigIds,
}) {
    return utils.runWithRetries(() => postAuth({
        url: '/suite/v2/list',
        body: {
            projectId,
            labels,
            testIds,
            testNames,
            testConfigNames,
            suiteNames,
            suiteIds,
            branch,
            rerunFailedByRunId,
            testConfigIds,
        },
    }));
}

function getUsageForCurrentBillingPeriod(projectId) {
    return utils.runWithRetries(() => getWithAuth(`/plan/project/${projectId}/usage-current-billing-period`))
        .catch((error) => {
            logger.error('failed getting usage for current billing period', { projectId, error });
            return undefined;
        });
}

function isTestResultCompleted(resultId, projectId, testRetryKey) {
    return utils.runWithRetries(() => getWithAuth(`/result/${resultId}/isComplete`, { projectId, testRetryKey }));
}

function getTestResults(testId, resultId, projectId, branch) {
    return utils.runWithRetries(() => getWithAuth(`/test/v2/${testId}/result/${resultId}`, { projectId, branch }));
}

function keepAliveGrid(projectId, slots) {
    return postAuth({
        url: `/grid/keep-alive?reqId=${utils.guid()}`,
        body: { projectId, slots },
        timeout: 10000,
    });
}

function releaseGridSlot(companyId, projectId, slotId, gridId, browser) {
    return postAuth({
        url: `/grid/release?reqId=${utils.guid()}`,
        body: {
            companyId, projectId, slotId, gridId, browser,
        },
    });
}

function getHybridGridProvider(body) {
    return postAuth({
        url: '/grid/hybrid/provider',
        body,
    });
}

function getGridByName(companyId, projectId, gridName, browser, executionId) {
    return utils.runWithRetries(() => getWithAuth('/grid/name', {
        companyId, projectId, name: gridName, browser, executionId, reqId: utils.guid(),
    }));
}

function getGridById(companyId, projectId, gridId, browser, executionId) {
    return utils.runWithRetries(() => getWithAuth(`/grid/${gridId}`, { companyId, projectId, browser, executionId, reqId: utils.guid() }));
}


async function initializeUserWithAuth({ projectId, token, branchName, lightweightMode, localGrid }) {
    try {
        return await utils.runWithRetries(() => httpRequest.post({
            url: `${config.SERVICES_HOST}/executions/initialize`,
            body: {
                projectId,
                token,
                branchName: branchName || 'master',
                lightweightMode,
                localGrid,
            },
        }));
    } catch (e) {
        logger.error('error initializing info from server', e);
        if (e && e.message && e.message.includes('Bad Request')) {
            throw new ArgError(
                'Error trying to retrieve CLI token. ' +
                'Your CLI token and project might not match. ' +
                'Please make sure to pass `--project` and `--token` that' +
                ' match to each other or make sure they match in your ~/.testim file.');
        }
        if (e && e.code && e.code.includes('ENOTFOUND')) {
            throw new ArgError('Due to network connectivity issues, Testim CLI has been unable to connect to the Testim backend.');
        }
        throw e;
    }
}

async function getEditorUrl() {
    if (config.EDITOR_URL) {
        return config.EDITOR_URL;
    }
    try {
        return await utils.runWithRetries(() => getWithAuth('/system-info/editor-url'));
    } catch (err) {
        logger.error('cannot retrieve editor-url from server');
        return 'https://app.testim.io';
    }
}

function getAllGrids(companyId) {
    return utils.runWithRetries(() => getWithAuth('/grid', { companyId }));
}

const fetchLambdatestConfig = async () => utils.runWithRetries(() => getWithAuth('/grid/lt/config'));

const getLabFeaturesByProjectId = async (projectId) => utils.runWithRetries(() => getWithAuth(`/labFeature/v2/project/${projectId}`));


function getRealData(projectId, channel, query) {
    return utils.runWithRetries(() => getWithAuth(`/real-data/${channel}?${query}&projectId=${projectId}`));
}

function updateTestResult(projectId, resultId, testId, testResult, remoteRunId) {
    return utils.runWithRetries(() => postAuth({
        url: '/result/test',
        body: {
            projectId,
            resultId,
            testId,
            testResult,
            remoteRunId,
        },
    }));
}

function clearTestResult(projectId, resultId, testId, testResult) {
    return utils.runWithRetries(() => postAuth({
        url: '/result/test/clear',
        body: {
            projectId,
            resultId,
            testId,
            testResult,
        },
    }));
}

function saveRemoteStep(projectId, resultId, stepId, remoteStep) {
    return utils.runWithRetries(() => postAuth({
        url: '/remoteStep',
        body: {
            projectId,
            resultId,
            stepId,
            remoteStep,
        },
    }));
}

function relativize(uri) {
    return uri.startsWith('/') ? uri : `/${uri}`;
}

function getStorageRelativePath(filePath, bucket, projectId) {
    let fullPath = relativize(filePath);
    if (projectId) {
        fullPath = `/${projectId}${fullPath}`;
    }
    if (bucket) {
        fullPath = `/${bucket}${fullPath}`;
    }

    return fullPath;
}

function uploadArtifact(projectId, testId, testResultId, content, subType, mimeType = 'application/octet-stream') {
    let fileSuffix = null;
    if (mimeType === 'application/json') {
        fileSuffix = '.json';
    }
    const fileName = `${subType}_${utils.guid()}${fileSuffix || ''}`;
    const path = `${testId}/${testResultId}/${fileName}`;
    const storagePath = getStorageRelativePath(path, 'test-result-artifacts', projectId);

    const buffer = Buffer.from(pako.gzip(content, {
        level: 3, // sufficient time/size ratio.
    }));

    const files = {
        file: {
            fileName,
            buffer,
        },
    };

    return utils.runWithRetries(() => postAuthFormData(`/storage${storagePath}`, {}, files, {
        'X-Asset-Encoding': 'gzip',
    })).then(() => storagePath);
}

const uploadRunDataArtifact = _.memoize(async (projectId, testId, testResultId, runData) => {
    if (_.isEmpty(runData)) {
        return undefined;
    }
    return await uploadArtifact(projectId, testId, testResultId, JSON.stringify(runData), 'test-run-data', 'application/json');
}, (projectId, testId, testResultId, runData) => `${hash(runData)}:${testId}:${testResultId}`);

const updateTestDataArtifact = _.memoize(async (projectId, testId, testResultId, testData, projectDefaults) => {
    if (_.isEmpty(testData)) {
        return undefined;
    }
    const removeHiddenParamsInTestData = () => {
        const testDataValueClone = _.clone(testData);
        if (projectDefaults && projectDefaults.hiddenParams) {
            const { hiddenParams } = projectDefaults;
            (hiddenParams || []).forEach((param) => {
                if (testDataValueClone[param]) {
                    testDataValueClone[param] = constants.test.HIDDEN_PARAM;
                }
            });
        }
        return testDataValueClone;
    };

    return await uploadArtifact(projectId, testId, testResultId, JSON.stringify(removeHiddenParamsInTestData(testData)), 'test-test-data', 'application/json');
}, (projectId, testId, testResultId, testData) => `${hash(testData)}:${testId}:${testResultId}`);

function addTestRetry({
    projectId,
    runId,
    testId,
    newResultId,
    originalTestResultId,
    previousTestResultId,
    testResult,
}) {
    return utils.runWithRetries(() => postAuth({
        url: '/result/test/retry',
        body: {
            projectId,
            newResultId,
            originalTestResultId,
            previousTestResultId,
            testId,
            runId,
            testResult,
        },
    }), DEFAULT_REQUEST_RETRY);
}

module.exports = {
    getS3Artifact,
    getTestPlan,
    saveTestPlanResult,
    updateTestStatus,
    updateExecutionTests,
    reportExecutionStarted,
    reportExecutionFinished,
    getTestPlanTestList,
    getSuiteTestList,
    getUsageForCurrentBillingPeriod,
    getTestResults,
    getGridByName,
    releaseGridSlot,
    keepAliveGrid,
    getGridById,
    getAllGrids,
    fetchLambdatestConfig,
    getRealData,
    updateTestResult,
    clearTestResult,
    saveRemoteStep,
    getEditorUrl,
    getLabFeaturesByProjectId,
    uploadRunDataArtifact: Promise.method(uploadRunDataArtifact),
    updateTestDataArtifact: Promise.method(updateTestDataArtifact),
    initializeUserWithAuth: Promise.method(initializeUserWithAuth),
    addTestRetry,
    getHybridGridProvider,
    loadTest,
    isTestResultCompleted,
};
